Allow request body content from file

warrenc5 opened the pull request:
I want to be able to have the webhook request body in a file in __files
e.g.
```
  "postServeActions": [
        {
            "name": "webhook",
            "parameters": {
                "method": "POST",
                "url": "http://localhost:8099/",
                "headers": {
                    "Content-Type": "application/json"
                },
                "bodyFileName": "req_scp_voice.json"
            }
        }
    ]

```

oleg-nenashev reviewed (commented):
The code looks good to me, but it would be nice to have some test coverage. Not sure about the docs, still new to the community

oleg-nenashev reviewed (approved):
This pull request looks good to me, thank you!
Since the REST API endpoint is updated, also a patch of the OpenAPI schema would be needed: https://github.com/wiremock/wiremock.org-sources/blob/6adf75b14e48fafdf50cbcca7d2e636eff87cbf0/swagger/schemas/response-definition.yaml . I guess raml too: https://github.com/wiremock/docs.wiremock.io-sources/tree/b833f8fee57189763b9c7f19181799ab70ef87db/assets/raml/schemas . Not sure whether they are codegenerated by now, would appreciate your feedback @tomakehurst

tomakehurst commented on wiremock-webhooks-extension/src/test/java/functional/WebhooksAcceptanceTest.java:
```diff
@@ -199,22 +202,30 @@ public void webhookCanBeConfiguredFromJson() throws Exception {
                             + "{{{jsonPath originalRequest.body '$.callbackPath'}}}")
                     .withHeader("X-Single", "{{math 1 '+' 2}}")
                     .withHeader("X-Multi", "{{math 3 'x' 2}}", "{{parameters.one}}")
                     .withBody("{{jsonPath originalRequest.body '$.name'}}")
```
I feel like specifying both the body and the body file is a bit confusing here. Perhaps 2 separate tests would be better rather than one parameterised one?

warrenc5 replied:
Yes - I have duplicated the tests now, one java/DSL set for bodyFileName and one set for body as a string.

oleg-nenashev reviewed (approved):
LGTM, v2.0

warrenc5 commented:
```
Execution failed for task ':wiremock-webhooks-extension:spotlessJavaCheck'.
> The following files had format violations:
```

But when I run that task on my local dev box I get 

```
gradlew :wiremock-webhooks-extension:spotlessJavaCheck

BUILD SUCCESSFUL in 343ms
3 actionable tasks: 3 up-to-date
```

Sorry, I'm unsure what the actual violation was at this point.

wirbe1 commented:
Very useful PR! 
Would be nice to have it in core webhooks functionality as well

leeturner commented:
Hi @warrenc5 it looks like this PR was pretty much over the line.  I think this would be a great addition to WireMock.  Are you interested in updating it with the latest `main` and we can get it over the line.  It looks like it was approved by Tom so we should all be good to go.  Are you still having issues with the spotless check?

Domadin commented:
@warrenc5 @leeturner 
Hello WireMock team! Is there an estimated timeline for merging this PR, which adds support for referencing files in webhook bodies? Our team has recently started using WireMock for stubs, and we’ve encountered the limitation of not being able to reference files in our webhook payloads. This functionality would greatly benefit our workflow. Could you please share any updates or guidance regarding this PR?
Thanks for your hard work!

maxsh8x commented:
@warrenc5 Could you please clarify — is the latest code from your master branch not working? When I try to use the feature, it attempts to parse bodyFileName as base64Body and crashes with an error "java.lang.IllegalArgumentException: Illegal base64 character 5f"

warrenc5 commented:
> @warrenc5 Could you please clarify — is the latest code from your master branch not working? When I try to use the feature, it attempts to parse bodyFileName as base64Body and crashes with an error "java.lang.IllegalArgumentException: Illegal base64 character 5f"

Thanks, I'll have a look - there's not enough info in your bug report to go on.

The crash error should be accompanied by a huge stacktrace giving the exact location of the exception. Can you include it here please?

Also, I can't see your project so I can't make any assumptions about what 5f actually means could you include a relevant snippet of that one also please?

maxsh8x commented:
> > @warrenc5 
> Thanks, I'll have a look - there's not enough info in your bug report to go on.

Good time of day. Here is one of the errors as an example. 

https://github.com/warrenc5/wiremock/blob/master/src/main/java/org/wiremock/webhooks/WebhookDefinition.java#L95

5th param - String bodyFileName

https://github.com/warrenc5/wiremock/blob/master/src/main/java/org/wiremock/webhooks/WebhookDefinition.java#L51

5th param - parameters.getString("base64Body", null)

warrenc5 commented:
Ok - thanks - good spotting @maxsh8x there's been a mixup of the invocation parameter order and the unit tests were not asserting the response body was the contents of the filename.

https://github.com/warrenc5/wiremock/issues/1

maxsh8x commented:
@warrenc5  Good time of day. I have checked the latest updates. 

In tests, everything might work correctly, but if you try to run the server in normal mode, `WireMockConfiguration.wireMockConfig().filesRoot();` will return an incorrect path for fileSource, something like `/src/test/resources/`, and we will get a "file not found" error. 

The correct path to filesSource can be obtained via `wireMockServices.getFiles()` This can be done approximately at this location: https://github.com/warrenc5/wiremock/blob/master/src/main/java/org/wiremock/webhooks/Webhooks.java#L59

warrenc5 commented:
Thanks @maxsh8x. Yes, You are correct. :)
There is some discrepancy around loading files from __files directory in test mode and standalone mode.

I also see that original request body paths in the postserve request template are not is not applied the same as parameters in the request from the tests. I don't know if this is by design yet.

I have created a standalone testcase to demonstrate the issue. 
I will push it shortly and continue to investigate.

warrenc5 commented:
I have created https://github.com/warrenc5/wiremock-test to test the post serve request body templating.

I have updated the test to use the __files child

warrenc5 commented:
From my perspective this change is ready to be integrated and I have no additional pending work on this PR. Please advise.

leeturner commented:
Could this PR be rebased with the new code in `main` ?  We have merged the 4.x beta code into the main branch now

warrenc5 commented:
Ok done, rebased against 4.0.0, double checked tests for postServe and serveEvents.
Updated standalone test case also https://github.com/warrenc5/wiremock-test

Also indirectly re-solved access to original response from webhook. https://github.com/wiremock/wiremock/pull/2838.

Please review and merge if all is well :)

warrenc5 commented:
Just waiting for @leeturner.
You could do your own build of this branch in the meanwhile, if that suits your use.

warrenc5 commented:
> Hello, any advancement on this PR ?

So since I started this pull request, some major relocations have happened core/common. 

It appears in my disoriented confusion I've been inconsistently merging/rebasing and basically polluting the commit log. :(

I am attempting to recreate this PR against a shiny new repo with a easy single squashed commit.

tomakehurst commented:
@warrenc5 I suggest not investing any more effort in this for the time being. We have an approach we've been discussing for a while to solve this problem in a way that works in other contexts e.g. binary matchers.

warrenc5 commented:
Moved to here https://github.com/wiremock/wiremock/pull/3278

tomakehurst commented:
I should have a branch to share later today with an example of this used in the websockets support I'm working on.

tomakehurst commented:
Bit later than anticipated, but here's an example of what I'm thinking:
https://github.com/wiremock/wiremock/blob/websockets-spike/wiremock-core/src/main/java/com/github/tomakehurst/wiremock/message/MessageDefinition.java#L27

The idea is that `EntityDefinition` will eventually be usable anywhere you want to define a body of some kind and might want to load it from the file system (or a custom store). The system will resolve this to an `Entity` before doing whatever work it needs with it.
